### PR TITLE
Stat stage related AI fixes

### PR DESCRIPTION
**Pull request by Sneed69**
## Issue(s) that this PR fixes
Added missing effect checks in `AI_IsMoveEffectInMinus`.
Some evasion and accuracy increasing effects were incorrectly handled.
Contrary self stat reduction checks were run but they were completely cancelled in `IncreaseStatUpScore` afterwards.

## **Discord contact info**
duke5614

### CODE REVIEW

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -3592,7 +3592,7 @@ static u32 AI_CalcMoveScore(u32 battlerAtk, u32 battlerDef, u32 move)
             else
                 ADJUST_SCORE(WEAK_EFFECT);
         }
-        else if (aiData->abilities[battlerAtk] != ABILITY_CONTRARY)
+        else
```
Comment (Sneed69) [inline src/battle_ai_main.c:3595]: This check was unnecessary as `IncreaseStatUpScore` already checks for contrary.

**Review by AlexOn1ine — CHANGES_REQUESTED** [review]
Good changes but can you please add 1-2 simple tests? It would be enough if you test something like Power Up Punch on Jirachi with Serene Grace. You can look into the `ai.c` file for examples.

**Comment by Sneed69** [comment]
Added more AI fixes, two tests and renamed `AI_CalcMoveScore` to `AI_CalcMoveEffectScore`.

File: src/battle_ai_util.c [hunk 2]
```diff
@@ -753,6 +770,25 @@ static bool32 AI_IsMoveEffectInMinus(u32 battlerAtk, u32 battlerDef, u32 move, s
...
+                case MOVE_EFFECT_ATK_PLUS_2:
+                case MOVE_EFFECT_DEF_PLUS_2:
+                case MOVE_EFFECT_SPD_PLUS_2:
+                case MOVE_EFFECT_SP_ATK_PLUS_2:
+                case MOVE_EFFECT_SP_DEF_PLUS_2:
+                case MOVE_EFFECT_EVS_PLUS_2:
+                case MOVE_EFFECT_ACC_PLUS_2:
+                case MOVE_EFFECT_ALL_STATS_UP:
+                    if ((gMovesInfo[move].additionalEffects[i].self && GetBattlerAbility(battlerAtk) == ABILITY_CONTRARY)
+                        || (noOfHitsToKo != 1 && abilityDef != ABILITY_CONTRARY && !IsMoldBreakerTypeAbility(abilityAtk)))
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:789]: I think the check for battleDef is unnecessary in this place. 

Also not related to the pr and can be changed later but for ai calcs it is better to use stored values so in this case `GetBattlerAbility(battlerAtk)` should be replaced with `AI_DATA->abilities{battlerAtk}`

Comment (Sneed69) [reply src/battle_ai_util.c:789]: I believe it's unnecessary right now, but necessary for custom moves or future moves.
The mold breaker check is a mistake however.

Comment (AlexOn1ine) [reply src/battle_ai_util.c:789]: could you give an example for a scenario where it could be needed?

Comment (Sneed69) [reply src/battle_ai_util.c]: For a move that deals damage and raises the target's attack.

Comment (AlexOn1ine) [reply src/battle_ai_util.c]: I see. That's fine then.

**Review by AlexOn1ine — DISMISSED** [review]

**Review by AlexOn1ine — APPROVED** [review]

**Pull request merged**